File: leetcode/recursion-backtracking/0079_word_search.py
# LC 79. Word Search | Medium
class Solution:
    def exist(self, board, word):
        m, n = len(board), len(board[0])
        def dfs(i, j, k):
            if k == len(word): return True
            if i < 0 or i >= m or j < 0 or j >= n or board[i][j] != word[k]: return False
            tmp, board[i][j] = board[i][j], '#'
            found = any(dfs(i+di, j+dj, k+1) for di, dj in [(0,1),(0,-1),(1,0),(-1,0)])
            board[i][j] = tmp; return found
        return any(dfs(i, j, 0) for i in range(m) for j in range(n))

File: leetcode/recursion-backtracking/0079_word_search.cpp
// LC 79. Word Search | Medium
#include <vector>
#include <string>
using namespace std;

class Solution {
    int m, n;
    bool dfs(vector<vector<char>>& b, string& w, int i, int j, int k) {
        if (k == w.size()) return true;
        if (i < 0 || i >= m || j < 0 || j >= n || b[i][j] != w[k]) return false;
        char tmp = b[i][j]; b[i][j] = '#';
        int d[] = {0,1,0,-1,0};
        for (int x = 0; x < 4; x++) if (dfs(b, w, i+d[x], j+d[x+1], k+1)) return true;
        b[i][j] = tmp; return false;
    }
public:
    bool exist(vector<vector<char>>& board, string word) {
        m = board.size(); n = board[0].size();
        for (int i = 0; i < m; i++) for (int j = 0; j < n; j++)
            if (dfs(board, word, i, j, 0)) return true;
        return false;
    }
};
